libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -452,25 +477,60 @@ string YulUtilFunctions::overflowCheckedIntDivFunction(IntegerType const& _type)
 				r := <?signed>s</signed>div(x, y)
 			}
 			)")
-				("functionName", functionName)
-				("signed", _type.isSigned())
-				("minVal", (0 - (u256(1) << (bits - 1))).str())
-				.render();
+			("functionName", functionName)
+			("signed", _type.isSigned())
+			("minVal", toCompactHexWithPrefix(u256(_type.minValue())))
+			.render();
 	});
 }
 
-string YulUtilFunctions::overflowCheckedUIntSubFunction()
+string YulUtilFunctions::checkedIntModFunction(IntegerType const& _type)
 {
-	string functionName = "checked_sub_uint";
+	string functionName = "checked_mod_" + _type.identifier();
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return
+			Whiskers(R"(
+			function <functionName>(x, y) -> r {
+				if iszero(y) { revert(0, 0) }
+				<?signed>
+				r := smod(x, y)
+				<!signed>
+				r := mod(x, y)
+				</signed>
+			}
+			)")
+			("functionName", functionName)
+			("signed", _type.isSigned())
+			.render();
+	});
+}
+
+string YulUtilFunctions::overflowCheckedIntSubFunction(IntegerType const& _type)
+{
+	string functionName = "checked_sub_" + _type.identifier();
 	return m_functionCollector->createFunction(functionName, [&] {
 		return
 			Whiskers(R"(
 			function <functionName>(x, y) -> diff {
+				<?signed>
+				)"
+				// underflow, if y >= 0 and x < (minValue - y)
```

Comment:
```suggestion
				// underflow, if y >= 0 and x < (maxValue - y)
```
(fixing myself, but please check)